test/tests/functional/pbs_job_array.py

```diff
@@ -744,3 +782,203 @@ def test_recover_big_array_job(self):
+        self.scheduler.log_match(errmsg, existence=False, max_attempts=10)
+        est = self.lu.convert_date_time(job_arr[0]['estimated.start_time'],
+                                        fmt="%a %b %d %H:%M:%S %Y")
+        self.assertAlmostEqual(stime + 300, est, 1)
+
+    @skipOnCpuSet
+    def test_max_run_subjobs_queuejob_hook(self):
+        """
+        Test that a queuejob hook is able to set max_run_subjobs attribute.
+        """
+        a = {'resources_available.ncpus': 8}
+        self.server.manager(MGR_CMD_SET, NODE, a, self.mom.shortname)
+
+        self.create_hook(3, "queuejob", "h1")
+        j1 = Job(attrs={ATTR_J: '1-20'})
+        jid1 = self.server.submit(j1)
+        self.server.log_match("max_run_subjobs set to 3")
+        self.server.expect(JOB, {ATTR_state: 'B'}, id=jid1)
+        self.server.expect(JOB, {'job_state=R': 3}, extend='t')
+
+    @skipOnCpuSet
+    def test_max_run_subjobs_modifyjob_hook(self):
+        """
+        Submit array job with large max_run_subjobs limit see if modifyjob
+        modifies it.
+        """
+        a = {'resources_available.ncpus': 20}
+        self.server.manager(MGR_CMD_SET, NODE, a, self.mom.shortname)
+
+        self.create_hook(3, "modifyjob", "h1")
+        self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': 'False'})
+        j = Job(attrs={ATTR_J: '1-50'})
+        jid = self.server.submit(j)
+        self.server.expect(JOB, {ATTR_state: 'Q'}, id=jid)
+        self.server.alterjob(jid, {ATTR_W: 'max_run_subjobs=20'})
+        self.server.log_match("max_run_subjobs set to 3")
+        self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': 'True'})
+        self.server.expect(JOB, {'job_state=R': 3}, extend='t')
+
+    @skipOnCpuSet
+    def test_max_run_subjobs_preemption(self):
+        """
+        Submit array job with max_run_subjobs limit and see if such a job
+        hits the limit, no preemption is attempted.
+        """
+        a = {'queue_type': 'execution',
+             'started': 'True',
+             'enabled': 'True',
+             'Priority': 200}
+        self.server.manager(MGR_CMD_CREATE, QUEUE, a, "wq2")
+
+        a = {'resources_available.ncpus': 8}
+        self.server.manager(MGR_CMD_SET, NODE, a, self.mom.shortname)
+
+        a = {'Resource_List.select': 'ncpus=2'}
+        j = Job(attrs=a)
+        jid = self.server.submit(j)
+        self.server.expect(JOB, {ATTR_state: 'R'}, id=jid)
+
+        a = {ATTR_J: '1-20%3', 'Resource_List.select': 'ncpus=2',
+             ATTR_q: 'wq2'}
+        j_arr = Job(attrs=a)
+        jid_arr = self.server.submit(j_arr)
+        self.server.expect(JOB, {ATTR_state: 'B'}, id=jid_arr)
+        self.server.expect(JOB, {'job_state=R': 4}, extend='t')
+        self.server.expect(JOB, {ATTR_state: 'R'}, id=jid)
+        msg = "Job cannot run due to max_run_subjobs limit,"
+        " preemption not attempted"
+        self.scheduler.log_match(jid_arr + ';' + msg)
```

Comment:
I don't think this test is necessary

Reply:
This test was to make sure that we don't try preemption when the job hits max_run_subjobs limit.

Reply:
I understand that.  Not trying preemption when the limit is hit is an optimization.  I wouldn't test for it.  Also, this message might go away when you move the code into the switch statement.

Reply:
Since this is a different kind of limit for which we have special code written to not try preemption, I thought we should add a test for it. I did remove the check for the log message while addressing your review comment.
I'd like to know what @tpai26 and @vstumpf think about it. If there is a consensus that the test is unnecessary, I'll remove it.